File: apps/mobile/src/services/historyService.ts
import AsyncStorage from '@react-native-async-storage/async-storage';

export interface HistoryItem {
  id: string;
  tool: string;
  query: string;
  result: string;
  timestamp: number;
  language: string;
}

const HISTORY_KEY = '@ai_tools_history';

export const historyService = {
  async saveItem(item: Omit<HistoryItem, 'id' | 'timestamp'>): Promise<void> {
    try {
      const existing = await this.getHistory();
      const newItem: HistoryItem = {
        ...item,
        id: Date.now().toString(),
        timestamp: Date.now(),
      };
      const updated = [newItem, ...existing];
      await AsyncStorage.setItem(HISTORY_KEY, JSON.stringify(updated));
    } catch (e) {
      console.error('Error saving history item', e);
    }
  },

  async getHistory(): Promise<HistoryItem[]> {
    try {
      const data = await AsyncStorage.getItem(HISTORY_KEY);
      return data ? JSON.parse(data) : [];
    } catch (e) {
      console.error('Error reading history', e);
      return [];
    }
  },

  async clearHistory(): Promise<void> {
    try {
      await AsyncStorage.removeItem(HISTORY_KEY);
    } catch (e) {
      console.error('Error clearing history', e);
    }
  },
  
  async deleteItem(id: string): Promise<void> {
    try {
      const existing = await this.getHistory();
      const updated = existing.filter(i => i.id !== id);
      await AsyncStorage.setItem(HISTORY_KEY, JSON.stringify(updated));
    } catch (e) {
      console.error('Error deleting history item', e);
    }
  },

  // Helper to group by "This Week", "Last Week", "Older"
  getGroupedHistory(items: HistoryItem[]) {
    const now = Date.now();
    const ONE_WEEK = 7 * 24 * 60 * 60 * 1000;
    const TWO_WEEKS = 14 * 24 * 60 * 60 * 1000;

    const groups = {
      'This Week': [] as HistoryItem[],
      'Last Week': [] as HistoryItem[],
      'Older': [] as HistoryItem[],
    };

    items.forEach(item => {
      const diff = now - item.timestamp;
      if (diff < ONE_WEEK) {
        groups['This Week'].push(item);
      } else if (diff < TWO_WEEKS) {
        groups['Last Week'].push(item);
      } else {
        groups['Older'].push(item);
      }
    });

    return [
      { title: 'This Week', data: groups['This Week'] },
      { title: 'Last Week', data: groups['Last Week'] },
      { title: 'Older', data: groups['Older'] }
    ].filter(g => g.data.length > 0);
  }
};

File: apps/supro_flutter/lib/features/ai_hub/services/history_service.dart
import 'dart:convert';
import 'package:shared_preferences/shared_preferences.dart';

class HistoryItem {
  final String id;
  final String tool;
  final String query;
  final String result;
  final int timestamp;
  final String language;

  HistoryItem({
    required this.id,
    required this.tool,
    required this.query,
    required this.result,
    required this.timestamp,
    required this.language,
  });

  Map<String, dynamic> toJson() => {
    'id': id,
    'tool': tool,
    'query': query,
    'result': result,
    'timestamp': timestamp,
    'language': language,
  };

  factory HistoryItem.fromJson(Map<String, dynamic> json) => HistoryItem(
    id: json['id'] ?? '',
    tool: json['tool'] ?? '',
    query: json['query'] ?? '',
    result: json['result'] ?? '',
    timestamp: json['timestamp'] ?? 0,
    language: json['language'] ?? 'Tamil',
  );
}

class HistoryGroup {
  final String title;
  final List<HistoryItem> data;
  HistoryGroup({required this.title, required this.data});
}

const String _historyKey = 'ai_tools_history';

class HistoryService {
  Future<void> saveItem({
    required String tool,
    required String query,
    required String result,
    required String language,
  }) async {
    try {
      final existing = await getHistory();
      final newItem = HistoryItem(
        id: DateTime.now().millisecondsSinceEpoch.toString(),
        tool: tool,
        query: query,
        result: result,
        timestamp: DateTime.now().millisecondsSinceEpoch,
        language: language,
      );
      final updated = [newItem, ...existing];
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(_historyKey, jsonEncode(updated.map((e) => e.toJson()).toList()));
    } catch (e) {
      // ignore
    }
  }

  Future<List<HistoryItem>> getHistory() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      final data = prefs.getString(_historyKey);
      if (data == null) return [];
      final list = jsonDecode(data) as List;
      return list.map((e) => HistoryItem.fromJson(e)).toList();
    } catch (e) {
      return [];
    }
  }

  Future<void> clearHistory() async {
    try {
      final prefs = await SharedPreferences.getInstance();
      await prefs.remove(_historyKey);
    } catch (e) {
      // ignore
    }
  }

  Future<void> deleteItem(String id) async {
    try {
      final existing = await getHistory();
      final updated = existing.where((i) => i.id != id).toList();
      final prefs = await SharedPreferences.getInstance();
      await prefs.setString(_historyKey, jsonEncode(updated.map((e) => e.toJson()).toList()));
    } catch (e) {
      // ignore
    }
  }

  List<HistoryGroup> getGroupedHistory(List<HistoryItem> items) {
    final now = DateTime.now().millisecondsSinceEpoch;
    const oneWeek = 7 * 24 * 60 * 60 * 1000;
    const twoWeeks = 14 * 24 * 60 * 60 * 1000;

    final thisWeek = <HistoryItem>[];
    final lastWeek = <HistoryItem>[];
    final older = <HistoryItem>[];

    for (final item in items) {
      final diff = now - item.timestamp;
      if (diff < oneWeek) {
        thisWeek.add(item);
      } else if (diff < twoWeeks) {
        lastWeek.add(item);
      } else {
        older.add(item);
      }
    }

    return [
      if (thisWeek.isNotEmpty) HistoryGroup(title: 'This Week', data: thisWeek),
      if (lastWeek.isNotEmpty) HistoryGroup(title: 'Last Week', data: lastWeek),
      if (older.isNotEmpty) HistoryGroup(title: 'Older', data: older),
    ];
  }
}
